File: Leetcode/src/com/leetcode/trie/_2416_SumofPrefixScoresofStrings.java
package com.leetcode.trie;

public class _2416_SumofPrefixScoresofStrings {
    class Solution {
        class TrieNode {
            TrieNode[] child;
            int count;
            
            public TrieNode() {
                child = new TrieNode[26];
                count = 0;
            }
        }
        // time: O(n * len(word))
        public int[] sumPrefixScores(String[] words) {
            // build the trie and keep counting
            if (words == null || words.length == 0) {
                return new int[]{};
            }
            TrieNode root = new TrieNode();
            for (String word : words) {
                TrieNode curr = root;
                for (char ch : word.toCharArray()) {
                    int index = ch - 'a';
                    if (curr.child[index] == null) {
                        curr.child[index] = new TrieNode();
                    }
                    // add count to next level
                    curr.child[index].count++;
                    curr = curr.child[index];
                }
            }
            
            int[] res = new int[words.length];
            for (int i = 0; i < words.length; i++) {
                TrieNode curr = root;
                for (char ch : words[i].toCharArray()) {
                    int index = ch - 'a';
                    res[i] += curr.child[index].count;
                    curr = curr.child[index];
                }
            }
            return res;
        }
    }
}

File: Leetcode/src/com/leetcode/trie/_2416_sum_of_prefix_scores_of_strings.py
class TrieNode:
    def __init__(self):
        self.children = dict()
        self.count = 0


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        curr = self.root
        for ch in word:
            if ch not in curr.children:
                curr.children[ch] = TrieNode()
            # whenever we travsal here
            curr.children[ch].count += 1
            curr = curr.children[ch]

    def compute_prefix_score(self, word: str):
        curr = self.root
        total = 0
        for ch in word:
            if ch not in curr.children:
                break
            total += curr.children[ch].count
            curr = curr.children[ch]
        return total


class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        t = Trie()
        for word in words:
            t.insert(word)

        return [t.compute_prefix_score(word) for word in words]
